src/main/kotlin/com/openlattice/data/storage/DataDeletionService.kt

```diff
@@ -135,43 +131,99 @@ class DataDeletionService(
+
+        /*
+        * 2 - Check authorization on all entities
+        */
+
+        /* Check entity set permissions */
+
+        val authorizedPropertiesByEntitySets = getAuthorizedPropertyTypesForDeleteByEntitySet(
+                setOf(entitySetId) + entitySetIdToEntityDataKeys.keys, // include original entity set too
+                deleteType,
+                Optional.empty(),
+                principals
+        )
+
+        /* Check association permissions */
+
+        // only original entity set can be association entity set
+        val isAssociation = entitySetManager.isAssociationEntitySet(entitySetId)
+        val authorizedAssociationPropertiesByEntitySets = if (isAssociation) {
+            // no need to check associations on original entity set
+            getAuthorizedPropertyTypesOfAssociations(entitySetIdToEntityDataKeys, deleteType, principals)
+        } else {
+            getAuthorizedPropertyTypesOfAssociations(
+                    mapOf(entitySetId to entityKeyIds) + entitySetIdToEntityDataKeys,
+                    deleteType,
+                    principals
+            )
+        }
+
 
         /*
-         * 2 - delete all entities
+         * 3 - Delete all entities
          */
 
         /* Delete entity */
 
         var numUpdates = 0
 
-        val entityWriteEvent = clearOrDeleteEntities(entitySetId, entityKeyIds, deleteType, principals)
+        val entityWriteEvent = clearOrDeleteAuthorizedEntities(
+                entitySetId,
+                entityKeyIds,
+                deleteType,
+                authorizedPropertiesByEntitySets.getValue(entitySetId)
+        )
         numUpdates += entityWriteEvent.numUpdates
 
-        /* 3 - Delete neighbors */
+        if (isAssociation) {
+            // if entity set is association, we only delete its edges
+            deleteEdgesForAssociationEntitySet(entitySetId, Optional.of(entityKeyIds), deleteType)
+        } else {
+            val associationNumUpdates = clearOrDeleteAuthorizedAssociations(
+                    entitySetId,
+                    Optional.of(entityKeyIds),
+                    deleteType,
+                    authorizedAssociationPropertiesByEntitySets.getValue(entitySetId)
+            )
+            numUpdates += associationNumUpdates
+        }
+
 
-        // TODO what is/was this for?
-        val neighborDeleteEvents = Lists.newArrayList<AuditableEvent>()
+        /* Delete neighbors */
 
-        numUpdates += entitySetIdToEntityDataKeysMap.entries.stream().mapToInt { entry ->
+        numUpdates += entitySetIdToEntityDataKeys.entries.stream().mapToInt { entry ->
             val neighborEntitySetId = entry.key
-            val neighborEntityKeyIds = entry.value.map { it.entityKeyId }.toSet()
+            val neighborEntityKeyIds = entry.value
 
-            val neighborWriteEvent = clearOrDeleteEntities(neighborEntitySetId,
+            val neighborWriteEvent = clearOrDeleteAuthorizedEntities(
```

Comment:
I could be reading this wrong, but I don't think you need to call `clearOrDeleteAuthorizedEntities` if you're also calling `clearOrDeleteAuthorizedAssociations`

Reply:
`clearOrDeleteAuthorizedAssociations` only deletes the connected association entities of the requested entities, so I think they both need to be called.. although in the opposite order, so the edges connected to entities won't be lost

Reply:
`clearOrDeleteAuthorizedAssociations` collects all connected edges first then it maps them to only contain the edge entityKeyIds from the DataEdgeKeys and then it deletes data, ids and edges of them